### PR TITLE
CMR-4134 - Add variable ingest routes and ingest functions

### PR DESCRIPTION
**Pull request by oubiwann**
This change does the following:
 * Creates routes for ingest
 * Creates ingest service functions in support of this
 * Reorganizes the ingest service functions (within the same namespace)

Additionally, it:
 * Splits the very large api.ingest namespace into functionally-related sub-namespaces 
 * Moved redundantly defined private function for extracting the user id from the context into common lib
 * This also required adding/updating messages namespaces
 * Does a little clean-up in other areas of the code, as I ran across things

What it doesn't do:
 * Test variable ingest update - I can remove the update code from this PR, if folks prefer
 * Validate JSON schemas (schema hasn't been finalized) - I can remove the validation support code from this branch, if desired

### CODE REVIEW

File: common-lib/src/cmr/common/api/context.clj [hunk 1]
```diff
@@ -48,3 +51,12 @@
   (when-let [request-id (context->request-id context)]
     {REQUEST_ID_HEADER request-id}))
 
+(defn context->user-id
```
Comment (ygliuvt) [inline common-lib/src/cmr/common/api/context.clj]: It's good that you are doing this. I almost did it myself, but decided to shelf it as I am not sure where to put this function. I have thought about putting it in `cmr.search.api.request-context-user-augmenter` as that is most closely related to the `util/lazy-assoc` and this function is only used by search-app. I would like to hear what other people think about it.

File: umm-spec-lib/src/cmr/umm_spec/umm_spec_core.clj [hunk 2]
```diff
@@ -24,14 +24,16 @@
    [cmr.umm-spec.xml-to-umm-mappings.iso19115-2 :as iso19115-2-to-umm]
    [cmr.umm-spec.xml-to-umm-mappings.serf :as serf-to-umm])
   (:import (cmr.umm_spec.models.umm_collection_models UMM-C)
```
Comment (ygliuvt) [inline umm-spec-lib/src/cmr/umm_spec/umm_spec_core.clj]: Put the namespace on the next line.

File: ingest-app/src/cmr/ingest/api/ingest/collections.clj [hunk 3]
```diff
@@ -0,0 +1,69 @@
+(ns cmr.ingest.api.ingest.collections
```
Comment (chris-durbin) [inline ingest-app/src/cmr/ingest/api/ingest/collections.clj]: Nit: Not sure I like the structure of ingest/api/ingest/... Why not put all of these refactored namespaces under ingest/api? I get that they were replacing a namespace called ingest.clj, but I think all of these are fine to group together at api.

File: ingest-app/src/cmr/ingest/api/ingest/routes.clj [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  operations."
+  (context "/variables" []
+    (POST "/"
+          {:keys [request-context headers body]}
+          (variables/create-variable request-context headers body))
+    (context "/:variable-key" [variable-key]
+      (PUT "/"
+           {:keys [request-context headers body]}
+           (variables/update-variable
+            request-context headers body variable-key)))))
```
Comment (chris-durbin) [inline ingest-app/src/cmr/ingest/api/ingest/routes.clj]: We should write a ticket for deleting variables. Are there any other tickets we need to write for the variables ingest API?

File: acl-lib/src/cmr/acl/core.clj [hunk 5]
```diff
@@ -159,14 +159,17 @@
...
-(defn- has-ingest-management-permission?
+(defn has-ingest-management-permission?
   "Returns true if the user identified by the token in the cache has been granted
   INGEST_MANAGEMENT_PERMISSION in ECHO ACLS for the given permission type."
   [context permission-type object-identity-type provider-id]
   ;; Performance optimization here of returning true if it's the system user.
   (or (transmit-config/echo-system-token? context)
       (let [acl-oit-key (echo-acls/acl-type->acl-key object-identity-type)]
-        (->> (get-permitting-acls context object-identity-type "INGEST_MANAGEMENT_ACL" permission-type)
+        (->> (get-permitting-acls context
```
Comment (oubiwann) [inline acl-lib/src/cmr/acl/core.clj]: I can't see long lines in my editor; I use two panes, each at 80 chars.

File: common-lib/src/cmr/common/api/context.clj (hunk 1 above)
Comment (oubiwann) [inline common-lib/src/cmr/common/api/context.clj]: Cool  :-) Yeah, wherever folks think this should live, that's where I'll put it; just say the word.

File: umm-spec-lib/src/cmr/umm_spec/umm_spec_core.clj (hunk 2 above)
Comment (oubiwann) [inline umm-spec-lib/src/cmr/umm_spec/umm_spec_core.clj]: Good eyes -- thanks.

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 6]
```diff
@@ -156,17 +158,32 @@
...
 
+(defn validate-variable-umm-spec-schema
+  "Validate the variable against the JSON schema and throw errors if
+  configured or return a list of warnings"
+  [variable validation-options]
+  (if-let [err-messages (seq (json-schema/validate-umm-json
+                              (umm-json/umm->json variable)
+                              :variable))]
+    (if (or (:validate-umm? validation-options)
+            (config/return-umm-json-validation-errors))
```
Comment (chris-durbin) [inline ingest-app/src/cmr/ingest/validation/validation.clj]: This was intended for UMM-C only. We'll need to introduce a separate config if we want to do something similar for variables. I don't think we have any need to do this though so we should remove the idea of warnings for variables.

File: system-int-test/src/cmr/system_int_test/utils/ingest_util.clj [hunk 7]
```diff
@@ -23,7 +23,8 @@
    [cmr.system-int-test.utils.dev-system-util :as dev-sys-util]
    [cmr.common-app.test.side-api :as side]
    [cmr.ingest.config :as icfg])
-  (:import [java.lang.NumberFormatException]))
+  (:import
+   [java.lang.NumberFormatException]))
```
Comment (chris-durbin) [inline system-int-test/src/cmr/system_int_test/utils/ingest_util.clj]: FYI - I think we agreed even for single imports to use the syntax:

`(java.lang NumberFormatException)`

File: system-int-test/src/cmr/system_int_test/utils/ingest_util.clj [hunk 8]
```diff
@@ -304,6 +320,56 @@
...
      (parse-ingest-response (client/request params) options))))
 
+(defn ingest-variable
+  "Ingest a variable."
+  ([variable]
+   (ingest-variable variable {}))
+  ([variable options]
+   (let [{:keys [token client-id user-id validate-keywords validate-umm-var cmr-request-id]} options
+         accept-format (:accept-format options)
+         headers (util/remove-nil-keys {"Cmr-Validate-Keywords" validate-keywords
```
Comment (chris-durbin) [inline system-int-test/src/cmr/system_int_test/utils/ingest_util.clj]: Similar comment validate keywords is specific to UMM-C, Validate-umm-var is probably unnecessary and should only be added if we have a good reason.

File: system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj [hunk 9]
```diff
@@ -0,0 +1,162 @@
+(ns cmr.system-int-test.ingest.variable-ingest-test
+  "CMR collection ingest integration tests"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj]: Update namespace docstring.

File: system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    [cmr.system-int-test.utils.search-util :as search]
+    [cmr.transmit.config :as transmit-config]
+    [cmr.umm-spec.models.umm-common-models :as umm-cmn]
+    [cmr.umm-spec.test.expected-conversion :as exc]
+    [cmr.umm-spec.test.location-keywords-helper :as lkt]
+    [cmr.umm-spec.umm-spec-core :as umm-spec]))
+
+(use-fixtures :each (ingest/reset-fixture))
+
+(defn- ingest-succeeded?
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj]: Function name is misleading - this is basically `has-permission-to-ingest`.

File: umm-spec-lib/src/cmr/umm_spec/umm_spec_core.clj (hunk 2 above)
Comment (oubiwann) [inline umm-spec-lib/src/cmr/umm_spec/umm_spec_core.clj]: Fixed.

File: ingest-app/src/cmr/ingest/api/ingest/bulk.clj [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  (:require
+   [clojure.data.xml :as xml]
+   [clojure.string :as string]
+   [cmr.acl.core :as acl]
+   [cmr.common.log :refer [debug info warn error]]
+   [cmr.common.mime-types :as mt]
+   [cmr.common.services.errors :as srvc-errors]
+   [cmr.common.xml.gen :refer :all]
+   [cmr.ingest.api.ingest.core :refer [
+     generate-ingest-response
```
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/bulk.clj]: Fixed.

File: ingest-app/src/cmr/ingest/api/ingest/routes.clj (hunk 4 above)
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/routes.clj]: Yeah, I think we should write an explicit ticket for updating variables. Most of the code that's in place should work, but that will give a nice place for doing the system integration tests.

File: system-int-test/src/cmr/system_int_test/utils/ingest_util.clj (hunk 7 above)
Comment (oubiwann) [inline system-int-test/src/cmr/system_int_test/utils/ingest_util.clj]: Yup, you're quite correct. I'll fix.

File: ingest-app/src/cmr/ingest/api/ingest/bulk.clj [hunk 12]
```diff
@@ -0,0 +1,122 @@
+(ns cmr.ingest.api.ingest.bulk
```
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/bulk.clj]: Yeah, I've been rebasing from master regularly -- will do again before merge. I'll ping @laurenfrederick here so she can give it a once-over and make sure nothing she's added has been lost in the merges/rebases.

File: ingest-app/src/cmr/ingest/api/ingest/core.clj [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+   [cmr.common.xml.gen :refer :all]
+   [cmr.ingest.services.ingest-service :as ingest]
+   [cmr.ingest.services.messages :as msg]
+   [cmr.ingest.services.providers-cache :as pc]
+   [cmr.transmit.config :as transmit-config]
+   [cmr.transmit.echo.tokens :as tokens])
+  (:import
+   (clojure.lang ExceptionInfo)))
+
+(def VALIDATE_KEYWORDS_HEADER "cmr-validate-keywords")
```
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/core.clj]: Moved.

File: ingest-app/src/cmr/ingest/api/ingest/collections.clj (hunk 3 above)
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/collections.clj]: Okay, that works for me.

Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/collections.clj]: Moved this code, but not yet pushed -- it's a fairly big shift, so I'm re-running tests before I push it up.

File: system-int-test/src/cmr/system_int_test/utils/ingest_util.clj (hunk 7 above)
Comment (oubiwann) [inline system-int-test/src/cmr/system_int_test/utils/ingest_util.clj]: Done.

File: ingest-app/src/cmr/ingest/api/ingest/collections.clj (hunk 3 above)
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/api/ingest/collections.clj]: Okay, this was a bigger change than expected, but everything's working again: tests pass.

File: ingest-app/src/cmr/ingest/validation/validation.clj (hunk 6 above)
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/validation/validation.clj]: Right; meant to replace this with just a simple schema validation ... which, it seems, is just `json-schema/validate-umm-json` ... so I'll delete this.

File: system-int-test/src/cmr/system_int_test/utils/ingest_util.clj (hunk 8 above)
Comment (oubiwann) [inline system-int-test/src/cmr/system_int_test/utils/ingest_util.clj]: Removed.

File: system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj (hunk 9 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj]: Done.

File: system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj (hunk 10 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/ingest/variable_ingest_test.clj]: Renamed.

**Comment by oubiwann** [comment]
All changes checked against ingest tests; rebased against master, will run full test suite, and then merge when that's good.

**Pull request merged**